linkis-engineconn-plugins/spark/pom.xml

```diff
@@ -202,13 +207,103 @@
       <artifactId>linkis-rpc</artifactId>
       <version>${project.version}</version>
     </dependency>
-
+    <dependency>
+      <groupId>org.apache.linkis</groupId>
+      <artifactId>linkis-hadoop-hdfs-client-shade</artifactId>
+      <version>${project.version}</version>
```

Comment:
move the spark-2.4-hadoop-3.3 profile to this pom,   and add the dependency to the profile

Reply:
yep, it is anther way, user need to compile with hadoop3 profile at the same time ,shade jar will be needed just in case.

Reply:
I have tried ,seems not as expectd.
for now i have excluded hdfs from spark-hive、spark-core.  and so on , if the ```linkis-hadoop-hdfs-client-shade ``` dependency moved to ```spark-2.4-hadoop-3.3``` profile, when we donot compile with hadoop-3.3 and spark-2.4-hadoop-3.3, the spark module will be lack of hdfs dependency, if we add linkis-hadoop-common to the spark module, the compile will be failed for ```Unrecognized Hadoop major version numb```. if we do not exclude hdfs from the spark-core, spark-hive and so on. when we compile with hadoop-3.3 and spark-2.4-hadoop-3.3, the spark module  output lib will have hadoop3 dependency,backing to our previous version problem.

Reply:
i have also tried to use relocation, but found spark have some force dependency for some hadoop-common class(shade class type will be rejected),  @jackxu2011 do you have better ideas ?
```
  def hadoopFile[K, V](path : scala.Predef.String, inputFormatClass : scala.Predef.Class[_ <: org.apache.hadoop.mapred.InputFormat[K, V]], keyClass : scala.Predef.Class[K], valueClass : scala.Predef.Class[V], minPartitions : scala.Int = { /* compiled code */ }) : org.apache.spark.rdd.RDD[scala.Tuple2[K, V]] = { /* compiled code */ }

```
```
  def transfer(sc: SparkContext, path: String, encoding: String): RDD[String] = {
    sc.hadoopFile(path, classOf[TextInputFormat], classOf[LongWritable], classOf[Text], 1)
      .map(p => new String(p._2.getBytes, 0, p._2.getLength, encoding))
  }
```

Reply:
> I have tried ,seems not as expectd. for now i have excluded hdfs from spark-hive、spark-core. and so on , if the `linkis-hadoop-hdfs-client-shade ` dependency moved to `spark-2.4-hadoop-3.3` profile, when we donot compile with hadoop-3.3 and spark-2.4-hadoop-3.3, the spark module will be lack of hdfs dependency, if we add linkis-hadoop-common to the spark module, the compile will be failed for `Unrecognized Hadoop major version numb`. if we do not exclude hdfs from the spark-core, spark-hive and so on. when we compile with hadoop-3.3 and spark-2.4-hadoop-3.3, the spark module output lib will have hadoop3 dependency,backing to our previous version problem.

linkis-hadoop-common  will in the classpath when run spark ec, because the linkis-hadoop-common is in lib/linkis-common/ public-module

Reply:
> i have also tried to use relocation, but found spark have some force dependency for some hadoop-common class(shade class type will be rejected), @jackxu2011 do you have better ideas ?
> 
> ```
>   def hadoopFile[K, V](path : scala.Predef.String, inputFormatClass : scala.Predef.Class[_ <: org.apache.hadoop.mapred.InputFormat[K, V]], keyClass : scala.Predef.Class[K], valueClass : scala.Predef.Class[V], minPartitions : scala.Int = { /* compiled code */ }) : org.apache.spark.rdd.RDD[scala.Tuple2[K, V]] = { /* compiled code */ }
> ```
> 
> ```
>   def transfer(sc: SparkContext, path: String, encoding: String): RDD[String] = {
>     sc.hadoopFile(path, classOf[TextInputFormat], classOf[LongWritable], classOf[Text], 1)
>       .map(p => new String(p._2.getBytes, 0, p._2.getLength, encoding))
>   }
> ```

if you rebuild the spark2.4 with hadoop3,  in my opinion,  the shade is useless? because the spark2.4 can directed run with hadoop3

Reply:
yep, currently when we run spark ec, hdfs dependency in ```linkis-hadoop-hdfs-client-shade``` will be effective ,though ```linkis-hadoop-common ``` in classpath

Reply:
> yep, currently when we run spark ec, hdfs dependency in `linkis-hadoop-hdfs-client-shade` will be effective ,though `linkis-hadoop-common ` in classpath

so, you can add `linkis-hadoop-common` as provided?

Reply:
> > yep, currently when we run spark ec, hdfs dependency in `linkis-hadoop-hdfs-client-shade` will be effective ,though `linkis-hadoop-common ` in classpath
> 
> so, you can add `linkis-hadoop-common` as provided?

em..,,not sure if the compile will pass ,i will try it later, ready to go company